extension-workspace-references.md

```diff
@@ -0,0 +1,115 @@
+
+The `workspace/xreferences` extension to the Language Server Protocol (LSP) enables a language server to export all of the references to dependencies that code inside of the workspace makes.
+
+Use case: clients of a language server can invoke `workspace/xreferences` in order to find references to dependencies. This information can then be stored inside of a database, which allows the caller to create a 'global mapping' of symbols in dependencies to the workspace they are used in (e.g. to see "how do other people use this symbol?").
+
+### Initialization
+
+`ServerCapabilities` may contain a new field to indicate server-side support for this extension:
+
+```typescript
+interface ServerCapabilities {
+  /* ... all fields from the base ServerCapabilities ... */
+
+  /**
+   * The server provides workspace references exporting support.
+   */
+  xworkspaceReferencesProvider?: boolean;
+}
+```
+
+#### Workspace References Request
+
+The workspace references request is sent from the client to the server to export project-wide references to dependencies. That is, the response strictly returns references in the project to symbols defined in dependencies.
+
+_Request_
+* method: 'workspace/xreferences'
+* params: `WorkspaceReferencesParams` defined as follows:
+```typescript
+/**
+ * The parameters of a Workspace References Request.
+ */
+interface WorkspaceReferencesParams {
+}
+```
+
+_Response_:
+* result: `ReferenceInformation[]` defined as follows:
+```typescript
+/**
+ * Represents information about a reference to programming constructs like
+ * variables, classes, interfaces etc.
+ */
+interface ReferenceInformation {
+	/**
+	 * The location in the workspace where the `symbol` has been referenced.
+	 */
+	reference: Location;
+
+	/**
+	 * Metadata information describing the symbol being referenced.
+	 */
+	symbol: ReferenceSymbolInformation;
+}
+```
+* error: code and message set in case an exception happens during the workspace references request.
+
+Where `ReferenceSymbolInformation` is defined as follows:
+
+```typescript
+/**
+ * Represents information about a programming construct like a variable, class,
+ * interface etc that has a reference to it. Effectively, it contains data similar
+ * to SymbolInformation except all fields are optional and a metadata field is
+ * present for language-specific data.
+ *
+ * ReferenceSymbolInformation does NOT always uniquely identify a symbol. The
+ * caller should effectively consider the returned information to be
+ * information about the symbol which _generally_ (but NOT always) identifies a
+ * single symbol.
+ */
+interface ReferenceSymbolInformation {
+    /**
+     * The name of this symbol (same as `SymbolInformation.name`).
+     */
+    name?: string;
+
+    /**
+     * The kind of this symbol (same as `SymbolInformation.kind`).
+     */
+    kind?: number;
+
+    /**
+     * The URI of this symbol (same as `SymbolInformation.location.uri`).
+     */
+    uri?: string;
```

Comment:
The motivation for not including a `range` is that it's actually not very useful for storage. Going back to the purpose of `workspace/xreferences`, the goal is to emit the references to dependencies and store them in a database. Code is often updated, and `range`s are ephemeral because of this. So, yes, I think we do know the range in all language servers and we are discarding it.

To offer a counterpoint to the inclusion of `range` -- if we had a `range` that we _could_ rely on after code changes, we would always know the definition location and could query any information about it on-the-fly. That would make the entire global refs DB in Sourcegraph essentially just `reference URI + range` -> `definition URI + range`. That would be awesome, but the range is ephemeral. So instead we use this approach in which the metadata about the definition _most of the time_ resolves to a single definition, and from that we could derive an exact location regardless of most basic code changes like lines shifting slightly around etc.